pkg/controllers/membercluster/v1beta1/membercluster_controller.go

```diff
@@ -169,15 +169,32 @@ func (r *Reconciler) handleDelete(ctx context.Context, mc *clusterv1beta1.Member
 		}
 		return runtime.Result{Requeue: true}, controller.NewUpdateIgnoreConflictError(r.updateMemberClusterStatus(ctx, mc))
 	}
+	if r.canForceDelete(mc, currentImc) {
+		klog.V(2).InfoS("Force delete the member cluster, by garbage collecting owned resources", "memberCluster", mcObjRef)
+		return runtime.Result{Requeue: true}, r.garbageCollect(ctx, mc)
+	}
 	klog.V(2).InfoS("Need to wait for the agent to leave", "memberCluster", mcObjRef, "joinedCondition", mcJoinedCondition)
 	// mark the imc as left to make sure the agent is leaving the fleet
 	if err := r.leave(ctx, mc, currentImc); err != nil {
 		klog.ErrorS(err, "Failed to mark the imc as leave", "memberCluster", mcObjRef)
 		return runtime.Result{}, err
 	}
 	// update the mc status to track the leaving status while we wait for all the agents to leave.
-	// once the imc is updated, the mc controller will reconcile again.
-	return runtime.Result{}, controller.NewUpdateIgnoreConflictError(r.updateMemberClusterStatus(ctx, mc))
+	// once the imc is updated, the mc controller will reconcile again ,or we reconcile to force delete
+	// the member cluster after force delete wait time.
+	return runtime.Result{RequeueAfter: r.ForceDeleteWaitTime}, controller.NewUpdateIgnoreConflictError(r.updateMemberClusterStatus(ctx, mc))
+}
+
+func (r *Reconciler) canForceDelete(mc *clusterv1beta1.MemberCluster, imc *clusterv1beta1.InternalMemberCluster) bool {
+	// we only force delete if member agent cannot reconcile an InternalMemberCluster in Leave state
+	// before the force delete wait time.
+	if imc.Spec.State != clusterv1beta1.ClusterStateLeave {
+		return false
+	}
+	if mc.DeletionTimestamp != nil && time.Since(mc.DeletionTimestamp.Time) < r.ForceDeleteWaitTime {
```

Comment:
when will the "mc.deletionTimeStamp != nil"

Reply:
That's just a nil check just to safe since we are accessing a pointer but as mentioned this is in the leave which meaning mc.DeletionTimestamp is non nil